jenkins-scripts/tools/outdated-job-runner.groovy

```diff
@@ -0,0 +1,223 @@
+    'ign_gui-ign-6-win',
+    'ignition_gui-ci-ign-gui6-homebrew-amd64',
+    'ignition_tools-ci-ign-tools1-homebrew-amd64',
+    'ignition_gazebo-ci-ign-gazebo3-focal-amd64',
+    'ignition_cmake-ci-ign-cmake2-focal-amd64',
+    'ign_sensors-gz-7-win',
+    'ignition_fuel-tools-ci-gz-fuel-tools8-focal-amd64',
+    'ignition_rendering-ci-ign-rendering6-homebrew-amd64',
+    'ign_msgs-ign-8-win',
+    'ignition_msgs-ci-gz-msgs9-focal-amd64',
+    'ignition_gazebo-ci-ign-gazebo6-homebrew-amd64',
+    'ign_gui-gz-7-win',
+    'sdformat-sdf-9-win',
+    'ignition_transport-ci-gz-transport12-homebrew-amd64',
+    'ignition_gazebo-ci-ign-gazebo6-focal-amd64',
+    'ignition_launch-ci-gz-launch6-focal-amd64',
+    'ignition_fuel-tools-ci-ign-fuel-tools4-focal-amd64',
+    'ign_cmake-gz-3-win',
+    'ign_gazebo-ign-6-win',
+    'ignition_sensors-ci-ign-sensors3-focal-amd64',
+    'sdformat-ci-sdformat13-homebrew-amd64',
+    'sdformat-ci-sdformat13-focal-amd64',
+    'ignition_common-ci-gz-common5-homebrew-amd64',
+    'ignition_cmake-ci-gz-cmake3-focal-amd64',
+    'ign_transport-ign-11-win',
+    'ign_gazebo-gz-7-win',
+    'ign_sensors-ign-6-win',
+    'ignition_sensors-ci-ign-sensors6-homebrew-amd64',
+    'ignition_gazebo-ci-gz-sim7-homebrew-amd64',
+    'ignition_gui-ci-gz-gui7-homebrew-amd64',
+    'ignition_math-ci-gz-math7-focal-amd64',
+    'ignition_transport-ci-ign-transport11-homebrew-amd64',
+    'ignition_math-ci-gz-math7-homebrew-amd64',
+    'ignition_fuel-tools-ci-ign-fuel-tools7-homebrew-amd64'
+];
+
+def trackedJobs = new LinkedHashSet(trackedJobsList);
+
+def runJob(String jobName) {
+    /*
+     * Schedule a job to run immediately
+     */
+    def job = Jenkins.instance.getItemByFullName(jobName)
+    job.scheduleBuild(0)
+}
+
+def findAvailableNodes() {
+    /*
+     * Find available nodes to run jobs per OS using labels
+     * win: windows
+     * osx: macos
+     * linux: docker
+     */
+    def availableNodes = [osx: [], win: [], linux: []]
+
+    Jenkins.instance.nodes.each { node ->
+        if (node.computer.online && node.computer.countIdle() > 0) {
+            if (node.getLabelString().contains("win")) {
+                availableNodes.win << node.name
+            } else if (node.getLabelString().contains("osx")) {
+                availableNodes.osx << node.name
+            } else if (node.getLabelString().contains("docker")) {
+                availableNodes.linux << node.name
+            }
+        }
+    }
+
+    return availableNodes
+}
+
+def runJobsInAvailableNodes(LinkedHashMap outdatedJobs) {
+    /*
+     * For each node in each OS, run a job if there are jobs to run
+     */
+    def availableNodes = findAvailableNodes()
+
+    availableNodes.each { os -> 
+        os.value.each { node ->
+                if (outdatedJobs[os.key].size() > 0) {
+                    def jobName = outdatedJobs[os.key].remove(0)
+                    println("Running job ${jobName} in node ${node}")
```

Comment:
I would say that there is no guarantee that the `jobName` is going to be run on the node `$node` unless we want to code that somehow here something that we probably don't want. So let's be humble in the output and say what is really happening: we are triggering a job :)

Reply:
Makes sense, and testing I found that some jobs aren't expecting `docker` label but `large-memory` label. Do you think I should categorize the jobs and nodes differently? Maybe pair job expected labels with node labels? :thinking:

Reply:
> Makes sense, and testing I found that some jobs aren't expecting `docker` label but `large-memory` label. Do you think I should categorize the jobs and nodes differently? Maybe pair job expected labels with node labels?

We also have the problem of the nodes requiring `gpu-nvidia` or `gpu-reliable`. Since all the linux nodes are tagged `docker` I think that to reduce the manual maintenance we can assume that all linux jobs are fine in any `docker` tagged node, even if that is not really true. Otherwise the manual maintenance in the list can be hard to manage. The worst case will be that we add a couple of jobs to the queue which does not seems extremely bad.